### DISCUSSION TITLE
Concatting middleware causes reducer type error

### DISCUSSION THREAD

**Discussion by chris-m92** [reduxjs/redux-toolkit#4272] [Q&A] [answered]
I'm trying to add a custom middleware, in essence to save the user's session via `localStorage` whenever an action is dispatched to my store.

So far I have the `saveSession.tsx` file:
```tsx
const saveSession = (store: any) => (next: any) => (action: any) => {
  console.log('Action: ', action)
  next(action)
}

export default saveSession
```

However, when attempting to concat this to the default middleware by

```tsx
export const store = configureStore({
  reducer: {
    entities: undoable(combineReducers({
      marker: markerStoreReducer,
      scenario: scenarioStoreReducer,
    }), {
      ...options
    }),
    ...moreReducers
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(saveSession)
```

I get the following TS error, specifically attached to the `entities` portion of my `reducer`
![image](https://github.com/reduxjs/redux-toolkit/assets/45608364/c66e55f8-a353-43b4-98ff-fd868df68ce5)

The weird thing is the site still runs in dev mode and adding a `//@ts-ignore` gets rid of the error, but what could be causing this issue?

If it helps, the `entities` are various `entityAdapters`

**Comment by markerikson** [comment, marked as answer]
What happens if you comment out the middleware line?

Conversely, what happens if you move the entire `undoable(otherReducers)` out to a separate variable, and then do `reducer: {entities: theUndoableReducer}}`?

**Reply by chris-m92** [reply]
Commenting out the middleware line removes the error. Even if I just do `middleware: (getDefaultMiddleware) => getDefaultMiddleware()` the error still exists. 

However, pulling the entity reducer to 
```tsx
const entityReducer = undoable(
  combineReducers(
    {
      marker: markerStoreReducer,
      scenario: scenarioStoreReducer
    }
  ),
  {
    ...options
  }
)
```

and doing 
```tsx
export const store = configureStore({
  reducer: {
    entities: entityReducer,
    data: dataStoreReducer,
    preference: preferenceStoreReducer,
    ui: uiStoreReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware()
})
```
seems to have done the trick.

Thank you so much!

**Reply by markerikson** [reply]
Yeah, we've seen cases where doing `reducer: someFunctionThatCreatesAReducer()` just doesn't work right with TS.  It seems to struggle carrying some of the types through properly.
